ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs

```diff
@@ -99,33 +100,68 @@ data CandidateState blk = CandidateState
       -- 'candidateChain'.
     }
 
+bracketChainSyncClient
+    :: ( MonadSTM m
+       , MonadThrow m
+       , Ord peer
+       )
+    => STM m (AnchoredFragment (Header blk))  -- ^ Get the current chain
+    -> STM m (ExtLedgerState blk)             -- ^ Get the current ledger state
+    -> TVar m (Map peer (TVar m (CandidateState blk)))
+       -- ^ The candidate chains, we need the whole map because we
+       -- (de)register nodes (@peer@).
+    -> peer
+    -> (TVar m (CandidateState blk) -> AnchoredFragment (Header blk) -> m a)
+    -> m a
+bracketChainSyncClient getCurrentChain getCurrentLedger
+                       varCandidates peer body =
+    bracket register unregister (uncurry body)
+  where
+    register = atomically $ do
+      curChain      <- getCurrentChain
+      curChainState <- ouroborosChainState <$> getCurrentLedger
+      -- We use our current chain, which contains the last @k@ headers, as
+      -- the initial chain for the candidate.
+      varCandidate <- newTVar CandidateState
+        { candidateChain       = curChain
+        , candidateChainState  = curChainState
+        }
+      modifyTVar' varCandidates $ Map.insert peer varCandidate
+      return (varCandidate, curChain)
+
+{-
+    unregister _ = atomically $
+      modifyTVar' varCandidates $ Map.delete peer
+-}
+    --FIXME: currently the chain sync client test relies on the variable still
+    -- being in the map after the sync client finishes. This is clearly wrong
+    -- but we need to fix that test before we can do this cleanup properly.
+    unregister _ = return ()
+
 -- | Chain sync client
 --
 -- This never terminates. In case of a failure, a 'ChainSyncClientException'
 -- is thrown. The network layer classifies exception such that the
 -- corresponding peer will never be chosen again.
 chainSyncClient
-    :: forall m peer blk.
+    :: forall m blk.
        ( MonadSTM m
        , MonadThrow (STM m)
        , ProtocolLedgerView blk
-       , Ord peer
        , Condense (Header blk)
        , Condense (ChainHash blk)
        )
     => Tracer m String
     -> NodeConfig (BlockProtocol blk)
     -> BlockchainTime m
     -> ClockSkew                                   -- ^ Maximum clock skew
-    -> STM m (AnchoredFragment (Header blk))       -- ^ Get the current chain
     -> STM m (ExtLedgerState blk)                  -- ^ Get the current ledger state
-    -> TVar m (Map peer (TVar m (CandidateState blk)))
-       -- ^ The candidate chains, we need the whole map because we
-       -- (de)register nodes (@peer@).
-    -> peer -> Consensus ChainSyncClient blk m
-chainSyncClient tracer cfg btime (ClockSkew maxSkew) getCurrentChain
-                getCurrentLedger varCandidates peer =
-    ChainSyncClient initialise
+    -> TVar m (CandidateState blk)                 -- ^ Our peer's state var
```

Comment:
It's actually nicer that the `TVar` is brought in scope here, now the local functions don't need to take it as an extra argument.